### PR TITLE
Fused elementwise layernorm

### PR DESCRIPTION
**Pull request by guangzlu**
Fused elementwise and normalization, which maybe a key module in some AI modules. (Such as add+layernorm is a key module in transformer)
Y = normalize(elementwise(A,B,...), gamma, beta)
Input matrixes have been made into tuple type.
Optimization in [PR#453](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/453) had been taken.
Writing and reading of Intermediate results had been reduced.
The performance has improved much compared with my former code in [PR#452](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/452) , as the table below shows. Tests were done with Rocm5.2.3 on 1 die of MI250. 
Current elapse time is the running time need with kernel in this PR , former elapse time is time need  with kernel in [PR#452](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/452) , unfused elapse time is time need when doing add and layernorm with ck separately. 

<html xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=PowerPoint.Slide>
<meta name=Generator content="Microsoft PowerPoint 15">
<style>

</style>
</head>

<body>

Matrix size ( M * N ) | Current elapse  time/ms | Former elapse  time/ms | Unfused elapse  time/ms | Performance Improvement
-- | -- | -- | -- | --
(48*256) * 1024 | 0.0744001 | 0.119536 | 0.142641 | 60.67%
(48*512) * 1024 | 0.141264 | 0.227729 | 0.274849 | 61.21%
(48*256) * 2048 | 0.140272 | 0.241329 | 0.286065 | 72.04%

</body>

</html>

Bandwidth: 
<html xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=PowerPoint.Slide>
<meta name=Generator content="Microsoft PowerPoint 15">
<style>

</style>
</head>

<body>

Matrix size ( M * N ) | bandwidth | Effiency   (Peak bw   on MI250 is 1.6TB/s)
-- | -- | --
(48*256) * 1024 | 945.110GB/s | 57.68%
(48*512) * 1024 | 995.504GB/s | 60.73%
(48*256) * 2048 | 1002.571GB/s | 61.16%

</body>

</html>

### CODE REVIEW

**Comment by rocking5566** [comment]
Add test in next PR

File: include/ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+                            I1);
+
+                        unpack2(elementwise_op, out_data_refs, in_data_refs);
+                    });
+                });
+                threadwise_welford.Run(c_thread_buf[iK0], mean_thread_buf, var_thread_buf);
+
+                if constexpr(!SweepOnce)
+                {
+                    threadwise_c_store.Run(thread_buffer_desc_m_k,
```
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp:382]: Why don't you write the result of elementwise into LDS?

File: include/ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+#include "ck/utility/data_type.hpp"
+#include "ck/tensor_operation/gpu/element/binary_element_wise_operation.hpp"
+#include "ck/tensor_operation/gpu/block/blockwise_welford.hpp"
+#include "ck/tensor_operation/gpu/thread/threadwise_welford.hpp"
+#include "ck/tensor_operation/gpu/thread/threadwise_tensor_slice_transfer.hpp"
+#include "ck/tensor_operation/gpu/element/element_wise_operation.hpp"
+
+namespace ck {
+
+// Y = LayerNorm(A + B, Beta, Gamma)
```
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp:15]: I thought the naming of ElementwiseOperation and AccElementwiseOperation is confusing.
device_gemm is a good example to imitate.
I suggest to rewrite the comment like this
X = A +B
Y = Layernorm(X, beta, gamme)
Therefore, you can rename 
ElementwiseOperation -> XElementwiseOperation
AccElementwiseOperation-> YElementwiseOperation

File: include/ck/tensor_operation/gpu/device/impl/device_elementwise_normalization_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
+    struct Argument : public BaseArgument
+    {
+        Argument(const std::vector<index_t> lengths,
+                 const std::array<std::vector<index_t>, NumInput> inStridesArray,
+                 const std::vector<index_t> gammaStrides,
+                 const std::vector<index_t> betaStrides,
+                 const std::vector<index_t> yStrides,
+                 const std::vector<index_t> reduceDims,
+                 ElementwiseOperation elementwise_op,
+                 AccElementwiseOperation acc_elementwise_op,
```
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/device/impl/device_elementwise_normalization_impl.hpp:273]: elementwise_op -> x_element_op
acc_elementwise_op -> y_element_op

File: include/ck/tensor_operation/gpu/device/impl/device_elementwise_normalization_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+
+#include "ck/tensor_operation/gpu/device/device_elementwise_normalization.hpp"
+#include "ck/tensor_operation/gpu/device/device_reduce.hpp"
+#include "ck/tensor_operation/gpu/device/impl/device_reduce_common.hpp"
+#include "ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp"
+#include "ck/tensor_operation/gpu/grid/gridwise_set_buffer_value.hpp"
+#include "ck/host_utility/device_prop.hpp"
+#include "ck/host_utility/kernel_launch.hpp"
+
+// Y = layernorm( A + B, Beta, Gamma )
```
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/device/impl/device_elementwise_normalization_impl.hpp:21]: You have more than one elementwise operation here.
I thought the naming of ElementwiseOperation and AccElementwiseOperation is confusing.
device_gemm is a good example to imitate.
I suggest to rewrite the comment like this
X = A +B
Y = Layernorm(X, beta, gamme)
Therefore, you can rename
ElementwiseOperation -> XElementwiseOperation
AccElementwiseOperation-> YElementwiseOperation

File: include/ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+template <typename InDataTypePointerTuple,
+          typename CDataType,
+          typename GammaDataType,
+          typename BetaDataType,
+          typename YDataType,
+          typename AccDataType,
+          typename ElementwiseOperation,
+          typename AccElementwiseOperation,
```
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp:23]: naming

File: include/ck/tensor_operation/gpu/device/impl/device_elementwise_normalization_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+          typename AccElementwiseOperation,
+          index_t Rank,               //
+          index_t NumReduceDim,       //
+          index_t BlockSize,          //
+          index_t MThreadClusterSize, // Num of threads in a block on M direction
+          index_t KThreadClusterSize, // Num of threads in a block on N direction
+          index_t MThreadSliceSize,   // Each thread calculate rows
+          index_t KThreadSliceSize,   // Each thread calculate columns
+          index_t XYSrcVectorDim,     // Dimension to do reduce
+          index_t XSrcVectorSize,     // Size to fetch source x
```
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/device/impl/device_elementwise_normalization_impl.hpp:87]: What is X here?
Do you mean InDataTypeTuple?
We need to define a consistent naming for input tensor tuple. (In is also OK)

**Comment by rocking5566** [comment]
Write the test like this
https://github.com/guangzlu/composable_kernel/blob/1b62bfaa2a42ed83da2692f6797a5f929c39946f/test/normalization/test_groupnorm_fp16.cpp

File: include/ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+                });
+                threadwise_welford.Run(c_thread_buf[iK0], mean_thread_buf, var_thread_buf);
+
+                if constexpr(!SweepOnce)
+                {
+                    threadwise_c_store.Run(thread_buffer_desc_m_k,
+                                           make_tuple(I0, I0),
+                                           c_thread_buf(iK0),
+                                           c_grid_desc_m_k,
+                                           c_global_val_buf);
```
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp:386]: Why don't you write the result of elementwise into LDS?
It is same as two kernels method for non Sweeponce case

File: include/ck/tensor_operation/gpu/device/impl/device_elementwise_normalization_impl.hpp (hunk 6 above)
Comment (guangzlu) [reply include/ck/tensor_operation/gpu/device/impl/device_elementwise_normalization_impl.hpp:87]: Yes, you are right, the naming is confused.
Really thank you for your advice.
I'll make it more reasonable.

File: include/ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp (hunk 7 above)
Comment (guangzlu) [reply include/ck/tensor_operation/gpu/grid/gridwise_elementwise_layernorm_welford_variance.hpp:386]: Yes, you are right, I'll change it, make the results into LDS.
Thanks for advice!

File: test/normalization/test_elementwise_layernorm_fp16.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    using BDataType     = std::tuple_element_t<1, Tuple>;
+    using GammaDataType = std::tuple_element_t<2, Tuple>;
+    using BetaDataType  = std::tuple_element_t<3, Tuple>;
+    using AccDataType   = std::tuple_element_t<4, Tuple>;
+    using YDataType     = std::tuple_element_t<5, Tuple>;
+
+    void Run()
+    {
+        // M, N
+        std::vector<std::vector<ck::index_t>> lengths = {{1024, 1024}};
```
Comment (rocking5566) [inline test/normalization/test_elementwise_layernorm_fp16.cpp:25]: Add some corner cases, eg: {1, 1}, {39, 777}

File: include/ck/tensor_operation/gpu/device/impl/device_elementwise_normalization_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,592 @@
...
+namespace ck {
+template <typename GridwiseElementwiseReduction,
+          typename InDataTypePointerTuple, // Datatype tuple of inputs
+          typename XDataType,              // Datatype of X
+          typename GammaDataType,          // Datatype of Gamma
+          typename BetaDataType,           // Datatype of Beta
+          typename YDataType,              // Datatype of input Y
+          typename AccDataType,            // AccDatatype
+          typename XElementwiseOperation,  // Operation of input
+          typename YElementwiseOperation,  // Operation Passthrough
```
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/device/impl/device_elementwise_normalization_impl.hpp:32]: Operation of Y

File: test/normalization/test_elementwise_layernorm_fp16.cpp (hunk 8 above)
Comment (rocking5566) [reply test/normalization/test_elementwise_layernorm_fp16.cpp]: Looks like you found the bugs from the corner cases. That is great!

**Review by rocking5566 — DISMISSED** [review]
LGTM. @asroy I thought this PR could be merged

**Comment by guangzlu** [comment]
> LGTM. @asroy I thought this PR could be merged

Really thanks for advices and reviewing!

File: test/elementwise_normalization/test_elementwise_layernorm_fp16.cpp [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+                                                                            YDataType>(
+                true, 2, false, false, length);
+            EXPECT_TRUE(success);
+        }
+    }
+};
+
+using KernelTypes = ::testing::Types<
+    // ADataType, BDataType, GammaDataType, BetaDataType, AccDataType, YDataType>
+    std::tuple<F16, F16, F16, F16, F32, F16>,
```
Comment (rocking5566) [inline test/elementwise_normalization/test_elementwise_layernorm_fp16.cpp:44]: Just need to leave one of them
